main/src/library/Iterator.flix

```diff
@@ -14,179 +14,158 @@
  * limitations under the License.
  */
 
-pub enum Iterator[a: Type, r: Region] {
-    case Iterator(Iterator.Done[r], Iterator.Next[a, r])
+pub enum Iterator[a: Type, ef: Bool, r: Region] {
+    case Iterator(Region[r], Unit -> Iterator.Step[a] \ {r, ef})
 }
 
+/*
 instance Newable[Iterator[a]] {
-    pub def new(r: Region[r]): Iterator[a, r] \ { Read(r), Write(r) } = Iterator.new(r)
-}
-
-instance Scoped[Iterator[a]] {
-    pub def regionOf(_: Iterator[a, r]): Region[r] = unsafe_cast () as Region[r]
+    pub def new(r: Region[r]): Iterator[a, ef, r] \ { Read(r), Write(r) } = Iterator.iterate(r)
 }
 
 ///
 /// Alias for `Foldable.iterator`.
 ///
-pub def iterator(r: Region[r], t: t[a]): Iterator[a, r] \ Write(r) with Foldable[t] =
+pub def iterator(r: Region[r], t: t[a]): Iterator[a, ef, r] \ Write(r) with Foldable[t] =
     Foldable.iterator(r, t)
+*/
 
 namespace Iterator {
 
-    ///
-    /// The type of the done function.
-    ///
-    pub type alias Done[ef: Region] = Unit -> Bool \ ef
+    pub enum Step[a] {
+        case Ans(a)
+        case Skip
+        case Done
+    }
 
     ///
-    /// The type of the next function.
+    /// Returns an iterator built with the stepper function `f`.
     ///
-    pub type alias Next[a: Type, ef: Region] = Unit -> a \ ef
+    pub def iterate(rc: Region[r], f: Unit -> Option[a] \ ef): Iterator[a, ef, r] =
+        let f1 = () -> match f() {
+            case Some(a) => Ans(a)
+            case None    => Done
+        };
+        let iterF = () -> unsafe_cast f1()  as _ \ { ef, r };
+        Iterator(rc, iterF)
 
     ///
     /// Returns an empty iterator.
     ///
-    pub def new(_: Region[r]): Iterator[a, r] \ { Read(r), Write(r) } =
-        let done = () -> unsafe_cast true as _ \ Read(r);
-        let next = () -> unsafe_cast bug!("Empty Iterator!") as _ \ Read(r);
-        unsafe_cast Iterator(done, next) as _ \ { Read(r), Write(r) }
+    pub def empty(rc: Region[r]): Iterator[a, ef, r] =
+        let f = () -> unsafe_cast Done as _ \ ef;
+        Iterator(rc, f)
```

Comment:
Definitely same issue as above.

Reply:
`empty` now casts to the Region `r`.

I've just noticed that `iterate` has become the odd one out of the "constructor" functions:

~~~

pub def iterate(rc: Region[r], f: Unit -> Option[a] \ ef): Iterator[a, ef, r]

~~~

For consistency it would have the following signature but that would contradict the type of `f` which doesn't necessarily access the region `rc` :

~~~

pub def iterate(rc: Region[r], f: Unit -> Option[a] \ ef): Iterator[a, r and ef, r]

~~~